libkbfs/cache/cache.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package cache
+
+import (
+	"sync"
+
+	"github.com/golang/groupcache/lru"
+)
+
+// Cache defines an interface for a cache that stores Measurable content.
+type Cache interface {
+	// Get tries to find and return data assiciated with key.
+	Get(key string) (data Measurable, ok bool)
+	// Add adds data into the cache, associating it with key. Entries are
+	// evicted when necessary.
+	Add(key string, data Measurable)
+}
+
+type randomEvictedCache struct {
+	maxBytes int
+
+	mu          sync.RWMutex
+	cachedBytes int
+	data        map[string]memorizedMeasurable
+}
+
+// NewRandomEvictedCache returns a Cache that uses random eviction strategy.
+// The cache will have a capacity of maxBytes bytes. A zero-byte capacity cache
+// is valid.
+//
+// Internally we store a memorizing wrapper for the raw Mesurable to avoid
+// unnecessarily frequent size calculations.
+func NewRandomEvictedCache(maxBytes int) Cache {
+	return &randomEvictedCache{
+		maxBytes: maxBytes,
+		data:     make(map[string]memorizedMeasurable),
+	}
+}
+
+// Get impelments the Cache interface.
+func (c *randomEvictedCache) Get(key string) (data Measurable, ok bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	memorized, ok := c.data[key]
+	if !ok {
+		return nil, false
+	}
+	return memorized.m, ok
+}
+
+// Add implements the Cache interface.
+func (c *randomEvictedCache) Add(key string, data Measurable) {
+	memorized := memorizedMeasurable{m: data}
+	if len(key)+memorized.Size() > c.maxBytes {
+		return
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if _, ok := c.data[key]; !ok {
+		c.cachedBytes += len(key) + memorized.Size()
+	}
+	if c.cachedBytes > c.maxBytes {
+		for k, v := range c.data {
```

Comment:
hmm somehow I thought that was in the spec. Thanks for bringing it up!

We could probably duplicate Go's `map` implementation and guarantee the random iteration ourselves, but that seems to be a pretty bad idea. I'll add comment for now.